src/main/webapp/app/exercises/programming/shared/code-editor/monaco/code-editor-monaco.component.ts

```diff
@@ -0,0 +1,217 @@
+     */
+    async onFileChange(fileChange: FileChange) {
+        if (fileChange instanceof RenameFileChange) {
+            this.fileSession = this.fileService.updateFileReferences(this.fileSession, fileChange);
+            for (const annotation of this.annotationsArray) {
+                if (annotation.fileName === fileChange.oldFileName) {
+                    annotation.fileName = fileChange.newFileName;
+                }
+            }
+            this.storeAnnotations([fileChange.newFileName]);
+        } else if (fileChange instanceof DeleteFileChange) {
+            this.fileSession = this.fileService.updateFileReferences(this.fileSession, fileChange);
+            this.storeAnnotations([fileChange.fileName]);
+        } else if (fileChange instanceof CreateFileChange && fileChange.fileType === FileType.FILE) {
+            this.fileSession = { ...this.fileSession, [fileChange.fileName]: { code: '', cursor: { row: 0, column: 0 }, loadingError: false } };
+        }
+        this.setBuildAnnotations(this.annotationsArray);
+    }
+
+    /*
+     * Taken from code-editor-ace.component.ts
+     */
+    /**
+     * Saves the updated annotations to local storage
+     * @param savedFiles
+     */
+    storeAnnotations(savedFiles: string[]) {
+        const toUpdate = fromPairs(this.annotationsArray.filter((a) => savedFiles.includes(a.fileName)).map((a) => [a.hash, a]));
+        const toKeep = pickBy(this.loadAnnotations(), (a) => !savedFiles.includes(a.fileName));
+
+        this.localStorageService.store(
+            'annotations-' + this.sessionId,
+            JSON.stringify({
+                ...toKeep,
+                ...toUpdate,
+            }),
+        );
+    }
+
+    /**
+     * Loads annotations from local storage
+     */
+    loadAnnotations() {
+        return JSON.parse(this.localStorageService.retrieve('annotations-' + this.sessionId) || '{}');
+    }
+
+    setBuildAnnotations(buildAnnotations: Annotation[]): void {
+        if (buildAnnotations.length > 0 && this.selectedFile) {
+            const sessionAnnotations = this.loadAnnotations();
+            this.annotationsArray = buildAnnotations.map((a) => {
+                const hash = a.fileName + a.row + a.column + a.text;
+                if (sessionAnnotations[hash] == undefined || sessionAnnotations[hash].timestamp < a.timestamp) {
+                    return { ...a, hash };
+                } else {
+                    return sessionAnnotations[hash];
+                }
+            });
+        } else {
+            this.annotationsArray = buildAnnotations;
+        }
+        this.editor.setAnnotations(
+            buildAnnotations.filter((buildAnnotation) => buildAnnotation.fileName === this.selectedFile),
+            this.commitState === CommitState.UNCOMMITTED_CHANGES,
+        );
+    }
+
+    // Expose to template
+    protected readonly Feedback = Feedback;
+    protected readonly CommitState = CommitState;
```

Comment:
We typically place these attributes at the top of the file (below the other "normal" attributes)

Reply:
Moved